src/psyclone/domain/gocean/kernel/psyir.py

```diff
@@ -222,16 +224,22 @@ def create_from_psyir(symbol):
 
         datatype = symbol.datatype
 
-        if not isinstance(datatype, UnsupportedFortranType):
+        if not isinstance(datatype, StructureType):
             raise InternalError(
                 f"Expected kernel metadata to be stored in the PSyIR as "
-                f"an UnsupportedFortranType, but found "
+                f"a StructureType, but found "
                 f"{type(datatype).__name__}.")
 
-        # In an UnsupportedFortranType, the declaration is stored as a
-        # string, so use create_from_fortran_string()
+        # TODO #2643: This is a temporary solution using FortranWriter
+        # to allow the current metadata extraction to work with StructureType,
+        # instead of relying on UnsupportedFortranType.
+        # This will be removed when the metadata is extracted from the PSyIR
+        # itself.
+        type_declaration = FortranWriter().gen_typedecl(symbol)
+        type_declaration = type_declaration.replace(", public", "")
+        type_declaration = type_declaration.replace(", private", "")
         return GOceanKernelMetadata.create_from_fortran_string(
```

Comment:
This would require reimplementing the PSyIR equivalent of the fparser2 based `GOceanKernelMetadata.create_from_fortran_string` and `GOceanKernelMetadata._get_property`, which I've for now left for #2643. It's clearly doable now, but I'd rather avoid doing this and leave the API/domain-specific edits for later, and probably to someone else :), unless you feel it's mandatory for this PR.

Reply:
Ah, I'd missed that this was GOcean. I agree, you can leave it.